packages/validation-manager/src/ValidationManager.ts

```diff
@@ -82,128 +102,157 @@ export class ValidationManager implements IValidationManager {
     }
   }
 
-  parseValidationResult (userOp: UserOperation, res: ValidationResultStructOutput): ValidationResult {
-    const mergedValidation = mergeValidationDataValues(res.returnInfo.accountValidationData, res.returnInfo.paymasterValidationData)
+  async getStakes (sender: string, paymaster?: string, factory?: string): Promise<{sender: StakeInfo, paymaster?: StakeInfo, factory?: StakeInfo}> {
+    const addrsArr = [sender, paymaster ?? AddressZero, factory ?? AddressZero]
+    const [ret] = await runContractScript(this.provider, new GetStakes__factory(), [this.entryPoint.address, addrsArr])
+    return {
+      sender: ret[0],
+      paymaster: paymaster != null ? ret[1] : null,
+      factory: factory != null ? ret[2] : null
+    }
+  }
 
-    function fillEntity (addr: string | undefined, info: StakeInfoStructOutput): StakeInfo | undefined {
-      if (addr == null || addr === AddressZero) return undefined
-      return {
-        addr,
-        stake: info.stake,
-        unstakeDelaySec: info.unstakeDelaySec
-      }
+  // generate validation result from trace: by decoding inner calls.
+  async generateValidationResult (op: UserOperation, tracerResult: ERC7562Call): Promise<ValidationResult> {
+    // const validationData = tracerResult.calls[0].output
+    const validateUserOpCallIndex = op.factory == null ? 0 : 1
```

Comment:
I would like to recommend that you add another utility function that turns the `tracerResult.calls` array to a named object like:
```
{
  senderCall: {},
  paymasterCall: {},
  factoryCall: null
}
```

It is taxing for the brain to deal with the code like ```const validateUserOpCallIndex = op.factory == null ? 0 : 1; const validatePaymasterCallIndex = validateUserOpCallIndex + 1```, what do you think?